### PR TITLE
Split instance state into Instance and VMM tables

### LINKED ISSUE #2315: Instances' sled assignments won't change if an instance is stopped and restarted
The Nexus external API's "instance start" command passes through to `Nexus::instance_start_runtime`:

https://github.com/oxidecomputer/omicron/blob/9d1bd5539225b6a225ba3655ceab852c71baa04a/nexus/src/external_api/http_entrypoints.rs#L3088

https://github.com/oxidecomputer/omicron/blob/9d1bd5539225b6a225ba3655ceab852c71baa04a/nexus/src/app/instance.rs#L358-L377

If I'm reading things right, this function selects the sled to which to send the runtime state update by looking at the instance record in CRDB without regard for the instance's current state:

https://github.com/oxidecomputer/omicron/blob/9d1bd5539225b6a225ba3655ceab852c71baa04a/nexus/src/app/instance.rs#L608-L619

This seems like the right thing to do if the instance is already incarnated on a sled somewhere. But if the instance is stopped and doesn't exist on any sled, this will try to create the instance on the sled on which it most recently ran, which might not have capacity for it (even though some other sled might). This function should distinguish the "instance already incarnated" and "instance stopped" cases and select a new sled in the latter case.

### LINKED ISSUE #2842: Instance Deletion saga looks destructive, even if instance should not be deleted
Let's suppose we have an instance in a state where it *should not* be deleted. Specifically, a case where `sid_delete_instance_record` , calling `project_delete_instance`, observes that the instance cannot be running.

This should be reproducible in a scenario where the instance is "running" - it's not an `ok_to_delete_instance_state`:

https://github.com/oxidecomputer/omicron/blob/e4a5dd0029946c0f7f56f4858475c2935e0eb9a9/nexus/db-queries/src/db/datastore/instance.rs#L228

If we execute the instance delete saga on this instance, we'll execute the following actions:

https://github.com/oxidecomputer/omicron/blob/e4a5dd0029946c0f7f56f4858475c2935e0eb9a9/nexus/src/app/sagas/instance_delete.rs#L69-L82

Here's what will happen:

- The V2P mappings will be deleted
- The ASIC configuration will be deleted
- The instance record *cannot* be deleted, because we aren't in a valid state. The saga will start unwinding from `instance_delete_record_action`.
- On the unwind path, the V2P mappings will *try* to be re-created via `sid_v2p_ensure_undo`

This is problematic for a couple reasons:

1. The (temporary) destruction of the V2P mappings are an observable side-effect of the failed instance deletion
2. As far as I can tell, the `sid_delete_network_config` function will delete all NAT mappings, and this destructive action will not be "undone"

This seems like it'll degrade the network functionality of the instance, even though it remains running.

It seems like we should delete the instance record **first**, before proceeding with the de-allocation of resources. This will validate the state of the instance before we actually perform destructive operations.

FYI @jmpesp , @internet-diglett .

### LINKED ISSUE #3139: Want all migration destination information in CRDB's instance table
The CRDB instance table records a migrating instance's destination Propolis ID, but not its destination sled or the destination Propolis's server IP. This is a minor hassle in a couple of places:

- When a migration succeeds, the source and target both know that migration has succeeded, but only the target can push a correct instance state update to Nexus (because the source doesn't have the right sled ID/Propolis IP to insert into the updated state).
- When a migration fails, Nexus can determine from the updates it receives from the relevant sleds that the destination Propolis is no longer needed and that its resources can be released. Ideally, before freeing resources, Nexus would explicitly verify with the destination sled that the destination Propolis really is gone, but Nexus doesn't know what the target sled was when migration fails (because there's no field for it).

### LINKED ISSUE #3228: instance creation path should consider reservoir size
#3200 added support in sled-agent to configure the VMM reservoir for guest memory. Currently, the reservoir is not used by default (#3223).

In order for the default use of the reservoir to be safe, the instance creation path must consider the last set reservoir size set by sled-agent when picking a sled. There is currently a [TODO](https://github.com/oxidecomputer/omicron/blob/490946c60b7c6690f0383ecd386eed82e7216c7a/nexus/db-queries/src/db/datastore/sled.rs#L138) to handle this.

### LINKED ISSUE #3877: instance create saga should produce useful log messages
In recent investigations of instance creation times for large instances, I found that there were nearly nothing in the nexus logs around instance creation, and was reliant only on propolis-server logs for details there. This was problematic in cases where the instance failed to be created (similar failure mode to #3417), because there was no propolis zone for me to inspect.

We should beef up the instance creation saga logging with an eye toward debuggability here.

### LINKED ISSUE #3883: Instance create with start = true should use start saga to start instances
#3873 created an instance start saga and began to use it for explicit instance start requests. The other way to start an instance is to set `start: true` in the instance creation request; in this case the instance create saga executes its own instance-start sequence. This should be reworked so that the create saga delegates to the new start saga.

### PR DESCRIPTION
**Pull request by gjcolombo**
**Reviewer note:** Regrettably but unavoidably, this PR is huge because so many things depend on the definition of an `Instance` in CRDB. I recommend reading individual commits and their descriptions to make things more manageable. The commits are arranged in more-or-less narrative order: database schema first, then supporting types, then sled agent, then Nexus DB queries, then Nexus instance functions, then sagas, then assorted bug fixes and bolt-tightening.

---

Refactor the definition of an `Instance` throughout the control plane so that an `Instance` is separate from the `Vmm`s that incarnate it. This confers several advantages:

- VMMs have their own state that sled agent can update without necessarily changing their instance's state. It's also possible to change an instance's active Propolis ID without having to know or update an instance's Propolis IP or current sled ID, since these change when an instance's active Propolis ID changes. This removes a great deal of complexity in sled agent, especially when live migrating an instance, and also simplifies the live migration saga considerably.
- Resource reservations for instances have much clearer lifetimes: a reservation can be released when its VMM has moved to a terminal state. Nexus no longer has to reason about VMM lifetimes from changes to an instance's Propolis ID columns.
- It's now possible for an Instance not to have an active Propolis at all! This allows an instance not to reserve sled resources when it's not running. It also allows an instance to stop and restart on a different sled.
- It's also possible to get a history of an instance's VMMs for, e.g., zone bundle examination purposes ("my VMM had a problem two days ago but it went away when I stopped and restarted it; can you investigate?").

Rework callers throughout Nexus who depend on knowing an instance's current state and/or its current sled ID. In many cases (e.g. disk and NIC attach and detach), the relevant detail is whether the instance has an active Propolis; for simplicity, augment these checks with "has an active Propolis ID" instead of trying to grab both instance and VMM states.

## Known issues/remaining work

- The virtual provisioning table is still updated only at instance creation/deletion time. Usage metrics that depend on this table might report strange and wonderful values if a user creates many more instances than can be started at one time.
- Instances still use the generic "resource attachment" CTE to manage attaching and detaching disks. Previously these queries looked at instance states; now they look at an instance's state and whether it has an active Propolis, but not at the active Propolis's state. This will need to be revisited in the future to support disk hotplug.
- `handle_instance_put_result` is still very aggressive about setting instances to the Failed state if sled agent returns errors other than invalid-request-flavored errors. I think we should reconsider this behavior, but this change is big enough as it is. I will file a TODO for this and update the new comments accordingly before this merges.
- The new live migration logic is not tested yet and differs from the "two-table" TLA+ model in RFD 361. More work will be needed here before we can declare live migration fully ready for selfhosting.
- It would be nice to have an `omdb vmm` command; for now I've just updated existing `omdb` commands to deal with the optionality of Propolises and sleds.

## Tests

- Unit/integration tests
- On a single-machine dev cluster, created two instances and verified that:
  - The instances only have resource reservations while they're running (and they reserve reservoir space now)
  - The instances can reach each other over their internal and external IPs when they're both running (and can still reach each other even if you try to delete one while it's active)
  - `scadm` shows the appropriate IP mappings being added/deleted as the instances start/stop
  - The instances' serial consoles work as expected
  - Attaching a new disk to an instance is only possible if the instance is stopped
  - Disk snapshot succeeds when invoked on a running instance's attached disk
  - Deleting an instance detaches its disks
- `omicron-stress` on a single-machine dev cluster ran for about an hour and created ~800 instances without any instances going to the Failed state (previously this would happen in the first 5-10 minutes)

## Related issues & PRs

Fixes #2315. Fixes #3139. Fixes #3883. Fixes #3228. Fixes #3877 (all the logic of interest from that issue is in the instance start saga now, and that saga has more verbose logging statements). Fixes #2842 (networking state is now torn down on instance stop). Mitigates #3325 (by making sled agent more careful to remove instances from its instance manager before transitioning them to stopped states). Supersedes PR #3141 (though see caveat above about how well live migration has--or hasn't--been tested here).

See also #3742.

### CODE REVIEW

**Comment by gjcolombo** [comment]
This is getting some cycles in the Canada region; once that comes back green (assuming it does...) I'll move this out of draft and assign reviewers.

File: nexus/src/app/sagas/instance_migrate.rs
Comment (smklein) [inline nexus/src/app/sagas/instance_migrate.rs:151]: (no action) this is a lot more readable than "reserve_on_random_sled ('please pick exactly this sled')" lol

File: nexus/src/app/sagas/instance_start.rs
Comment (smklein) [inline nexus/src/app/sagas/instance_start.rs:97]: (Note: I do not want to change this now, but asking out of curiosity)

From your "Known Issues" in the PR message:

> The virtual provisioning table is still updated only at instance creation/deletion time. Usage metrics that depend on this table might report strange and wonderful values if a user creates many more instances than can be started at one time.

We could basically "just" move the virtual provisioning saga step from instance creation here (and the corresponding release from "instance delete" to "instance stop"), right? Do you foresee anything being particularly more complex about this, after this PR merges?

I'm not even sure we **should** do so, but asking mostly to understand what we could do.

Comment (smklein) [reply nexus/src/app/sagas/instance_start.rs:97]: Follow-up -- I see some comments below about safely unwinding this saga if it is called concurrently with distinct propolis IDs. I see that would probably also be an additional refactor necessary for correctness -- the "propolis ID"s would need to consume virtual resources, rather than instance IDs, otherwise unwinding them wouldn't be safe.

File: sled-agent/src/common/instance.rs
Comment (smklein) [inline sled-agent/src/common/instance.rs:433]: One other idea for consideration, if we consider this whole aspect of "sled agent destroyed dead propolis" a risk:

We could punt it to Nexus? Specifically, rather than than Sled Agents autonomously destroying VMMs / their zones, they could set a flag, signal to Nexus that something bad has happened, and let Nexus be responsible for sending a follow-up message to destroy the zone. It's a bit more of a round-trip, but it seems like (1) it would give Nexus a chance to avoid some of the race conditions you've observed, and (2) it would give us a better-defined period of time to extract whatever info we want from the zone before deleting it.

All that said, this is an idea, I don't think it should block the PR. But I noticed the set of "autonomous sled agent actions performed in response to propolis" has whittled down to only "Destroy", so removing that last one might make data flow a bit easier.

**Review by smklein — COMMENTED** [review]
This PR is a colossal achievement - these changes make a ton of sense, and 100% push the stack in the right direction. Really looking forward to getting this merged, and getting mileage on it ASAP.

File: schema/crdb/README.adoc
Comment (gjcolombo) [reply schema/crdb/README.adoc:125]: IIRC (I don't have this in my notes, sadly) this doesn't quite work--my recollection is that it changes the opaque column ID and so changes the generated CHECK constraint name. I could try it again in a separate (smaller) branch to see what happens.

File: sled-agent/src/common/instance.rs
Comment (gjcolombo) [reply sled-agent/src/common/instance.rs:433]: I think I had a somewhat different case in mind when I wrote this comment: suppose we finally have a "force-quit this instance" API, and you call it on a migrating instance, and it starts by terminating the current active Propolis (i.e. the migration source). Then you could hit the race described here--the instance generation numbers are

Thinking about that some more, though, I think I would solve it differently, essentially by having some kind of rule that says something like "instance updates from sled agent are only applicable if the instance still points to the associated VMM ID." Then Nexus can "disown" all the running VMMs in a single transaction (by nulling out the UUIDs and writing an instance state that forbids the instance from restarting) and then make sure the individual VMMs have stopped. (You'd probably have to do this in a saga.)

Another approach I've considered (but not worked out in detail) is related to the "who owns the fallback state?" question you raised in other comments: instead of having sled agent send modified Instance records to Nexus, have it send more abstract messages about what a VMM has done ("VMM says it migrated out," "VMM went to the Great Control Plane in the Sky, etc.") and let Nexus own all the logic that touches Instance records. This would neatly answer the question about who owns the Instance state field (it's in the Instance so Nexus does!) but we then have to reason about, say, parallel messages about the same instance arriving from different sleds at the same time. Aren't distributed systems fun?

---

With respect to instances that actually fail (instead of being force-stopped from the API): I want to more or less completely revisit how we use the Failed state today and how we distinguish cases like "the sled reported this VMM had a problem" from "uh-oh, I couldn't talk to the sled, maybe there is a problem?" I think what you're describing here might fall into that. This ties into the comments in `handle_instance_put_result` for which I need to file an issue to get a number to put into the TODOs.

File: nexus/src/app/sagas/instance_migrate.rs
Comment (gjcolombo) [reply nexus/src/app/sagas/instance_migrate.rs:151]: I mean, we still do that. You just don't have to think about it unless you go into that specific file ;)

File: nexus/src/app/sagas/instance_delete.rs
Comment (gjcolombo) [reply nexus/src/app/sagas/instance_delete.rs:63]: Correct, this is (or should be) done during instance stop now, and the saga enforces in part that the instance is stopped by verifying that it has no active VMM in the database. (Assuming I understand #2842's root cause correctly, this is the part of the change that fixes it.)

File: nexus/src/app/instance.rs
Comment (gjcolombo) [reply nexus/src/app/instance.rs:257]: It's not totally infallible; you could still fail to create the VM. But otherwise you're right: in the old code, if you said `start: true` and the VM failed to start, the whole saga would unwind and the created VM would be destroyed. Now the behavior matches what you'd get if you passed `start: false` and then issued a separate `instance start` command once the VM was successfully created. I happen to like that symmetry, but don't feel especially strongly about it.

If (or when) we want the old behavior back, we can get it by trying to turn the start saga into a subsaga of the create saga. I don't think that's impossible, but there's a mild impedance mismatch in there that we'd have to resolve: the start saga expects to get a `db::Instance` as a parameter; to use it as a subsaga, the create saga's `make_saga_dag` has to be able to produce the new instance record in advance; I think that might be possible in principle, but we'd have to rework the create saga to do that, since it currently creates the instance record inline.

File: nexus/db-queries/src/db/queries/network_interface.rs
Comment (gjcolombo) [reply nexus/db-queries/src/db/queries/network_interface.rs:1290]: I think that makes more sense than leaving it in Starting like we do today. I'll see if I can work that into the start saga somewhere.

File: nexus/db-queries/src/db/queries/instance.rs
Comment (gjcolombo) [reply nexus/db-queries/src/db/queries/instance.rs:45]: Honestly, the hardest part was figuring out that I could do the `SELECT(SELECT(` trick to avoid having the query fail if one of the relevant objects was garbage-collected...

I'll give RFD 436 a look and see if I can leave a helpful comment about this.

File: nexus/db-queries/src/db/queries/instance.rs
Comment (gjcolombo) [reply nexus/db-queries/src/db/queries/instance.rs:220]: Good call, will fix.

File: nexus/db-queries/src/db/queries/instance.rs
Comment (gjcolombo) [reply nexus/db-queries/src/db/queries/instance.rs:224]: Good call, will fix this too.

File: nexus/db-queries/src/db/datastore/instance.rs
Comment (gjcolombo) [reply nexus/db-queries/src/db/datastore/instance.rs:297]: Yeah, I'm really glad not to have to concede to "we can only have one table!" anymore :)

File: nexus/db-model/src/vmm.rs
Comment (gjcolombo) [reply nexus/db-model/src/vmm.rs:13]: I think this is a mistake in the API type definition. `omicron_common::api::internal::nexus::VmmRuntimeState` is only supposed to contain the mutable bits of VMM runtime state. I probably got sloppy with defining all these types in several different places. Will try to get this cleaned up. Thanks for the catch!

File: nexus/db-model/src/schema.rs
Comment (gjcolombo) [reply nexus/db-model/src/schema.rs:369]: I was pattern-matching on what's in the instance (which has this and a `time_modified` field, which makes sense to me in that you can mutate the instance's fixed properties without mutating its state). I'm happy to change this if you feel strongly about it--let me know.

File: nexus/db-model/src/instance.rs
Comment (gjcolombo) [reply nexus/db-model/src/instance.rs:136]: Yeah... I wanted to rename this, but ran into the "anonymous CHECK constraints" problem we discussed elsewhere. If we can find a good way out of that I'd love to have a better column name here.

File: nexus/db-model/src/instance.rs
Comment (gjcolombo) [reply nexus/db-model/src/instance.rs:60]: Good idea. I'll do something similar to what I did for new VMM records.

File: common/src/api/internal/nexus.rs
Comment (gjcolombo) [reply common/src/api/internal/nexus.rs:49]: We talked about this some offline (and I discussed it some in another comment). I think the ideal semantics are

- when there's an VMM, sled agent owns the instance's logical state and puts it in the active VMM's record
- when there's no VMM, Nexus owns the instance's logical state and puts it in the instance record

Right now this breaks down a bit because sled agent updates the fallback state when an active Propolis is retired. We could change that a bit--if Nexus sees that the active Propolis has gone away, then it can insert the appropriate state into the instance record before writing anything back to CRDB.

I'm going to think about this one a bit more/sleep on it and then decide how to move forward.

File: nexus/tests/integration_tests/schema.rs
Comment (gjcolombo) [reply nexus/tests/integration_tests/schema.rs:539]: Short answer: yep.

Long answer: The backstory here is that I ran into the anonymous-CHECK-constraints problem a couple of weeks ago (before the change to add the table constraints check landed), and I had trouble diagnosing it because it was impossible to tell which tables were being assigned which opaque IDs. At that point I added my own table constraints check, which helped to make the problem more obvious (the errors had real table names in them!), and then I reordered everything to try to make the most obvious/easiest-to-diagnose failures come first.

File: schema/crdb/6.0.0/up.sql
Comment (gjcolombo) [reply schema/crdb/6.0.0/up.sql:1]: Oops, I waltzed past this part of the docs. Will figure out how to rearrange this. Thanks for the catch!

**Review by gjcolombo — COMMENTED** [review]
Thanks for the thorough feedback! I'll start resolving the easy stuff tomorrow and think some about the harder questions.

File: nexus/src/app/sagas/instance_start.rs
Comment (gjcolombo) [reply nexus/src/app/sagas/instance_start.rs:97]: Is this the same problem as we had in #2839? Seems like it wouldn't be too hard to resolve by a similar method if we wanted (use Propolis IDs instead of instance IDs, maybe change an enum variant or two around to reflect the sort of thing that's reserving the resources).

File: dev-tools/omdb/src/bin/omdb/db.rs
Comment (gjcolombo) [reply dev-tools/omdb/src/bin/omdb/db.rs:81]: I put it to a vote:

```
greg@augustine [Fri Oct 06 19:31:44] ~/omicron (gjcolombo/your-vmm-table-is-ready)
$ rg vmm_id | wc -l
      54
greg@augustine [Fri Oct 06 19:31:46] ~/omicron (gjcolombo/your-vmm-table-is-ready)
$ rg propolis_id | wc -l
     211
```

Well then! I mostly standardized on `propolis_id` in 78f72e5ce (with the exception of datastore code where `vmm_id` very specifically means "a primary key in the VMM table").

**Review by gjcolombo — COMMENTED** [review]
I've fixed a bunch of the easier stuff but haven't gotten around to the `fallback_state` thing yet.

File: nexus/db-model/src/instance.rs
Comment (gjcolombo) [reply nexus/db-model/src/instance.rs]: This function had only two callers and they both passed Creating (contrast with creating a new VMM where the initial state might be either Starting or Migrating), so I just got rid of the parameter entirely in f20a6d8.

File: common/src/api/internal/nexus.rs
Comment (gjcolombo) [reply common/src/api/internal/nexus.rs]: Fixed in d2a063668: the `fallback_state` field is now named `nexus_state`, and Nexus is the only thing that modifies it (enforced by removing the corresponding field from the structs Nexus and sled agent send each other). The instance state machine should now reduce to the following:

- Creating: this instance record is still being set up
- Stopped: this instance has no active VMM and so isn't running anywhere
- Failed: similar to Stopped except that Nexus thinks the instance had a Problem
- Running: the instance has an active VMM; go look at it for its state
- Destroyed: the instance is gone (it should also have a non-NULL `time_deleted`)

Instead of getting the Running -> Stopped transition from sled agent, Nexus recomputes the effective state when applying sled agent state updates based on whether the updated runtime states have a `propolis_id` or not. (This is the computation sled agent was doing; it's just happening up in Nexus now.)

This was a surprisingly small change (35 added, 68 removed, and a lot of that was just renaming), which makes me more confident it's correct. I also audited the code for places where we assign the Nexus state and verified that only the states I listed above are actually used.

> Does it make sense to have this be a more-constrained set of states than InstanceState?

Yes, we should enforce this whole thing by constraining the values this column can have. I'd like to do that in a separate PR because I expect it to be more tedious (we'll need to figure out the schema update and then create/plumb in all the new Rust types).

File: schema/crdb/README.adoc
Comment (gjcolombo) [reply schema/crdb/README.adoc]: I amended the docs in b8cbde5e1 to suggest this, but would suggest deferring enforcement to another PR.

File: nexus/db-queries/src/db/datastore/instance.rs
Comment (gjcolombo) [reply nexus/db-queries/src/db/datastore/instance.rs:348]: Touched up the comment in 8923fb4dc.

Updating both tables in a single update is by design. Sled agent updates its instance/VMM state pairs atomically (under a lock that protects each `InstanceStates` structure), and I didn't want to have to make sled agent reason about the order in which Nexus would apply updates that sled agent computed atomically and sent to Nexus in a single unit.

This is definitely meant for the "sled agent sent me an update" case, though--Nexus functions that aren't dealing directly with sled agent generally don't want to do this. I updated the doc comment to mention this. I'm not sure if there's a good way to give that requirement more teeth through the module system and am definitely open to suggestions there.

**Comment by gjcolombo** [comment]
@smklein This is ready for another lap at your convenience. I've resolved all the comment threads that needed no action or where I made a code change; there are a few things where there's still some discussion to be had and I've not made any changes yet, and I've left those open.

Thank you again for wrestling with this giant change!

**Review by jordanhendricks — APPROVED** [review]
I gave this a look through, mostly for my own understanding; I don't have a lot of useful feedback. Thanks for taking this on -- this is a huge change and I know it was a ton of work! Glad to see it land.

File: nexus/src/app/instance.rs
Comment (gjcolombo) [reply nexus/src/app/instance.rs:62]: This should read something closer to "the sled agent containing the instance's current VMM, as identified by its `propolis_id` field." I've updated the comment in 2933c4b9a to make this more explicit (will be part of my next push).

File: schema/crdb/README.adoc
Comment (smklein) [reply schema/crdb/README.adoc:130]: Right, this does seem conditional on having the constraints be explicitly named.

I filed https://github.com/oxidecomputer/omicron/issues/4256 to track, for later

File: common/src/api/internal/nexus.rs
Comment (smklein) [reply common/src/api/internal/nexus.rs]: This is great, thank you for taking the time to do this pass. I like this new version a lot!

**Review by smklein — APPROVED** [review]
Thanks for the follow-up on the `fallback_state` and SQL schema change stuff.

Let's get this in!

File: nexus/src/app/sagas/instance_start.rs
Comment (gjcolombo) [reply nexus/src/app/sagas/instance_start.rs:97]: Filed #4257 to track this.

**Pull request merged**
